### DISCUSSION TITLE
sdm on Mac OS Catalina/Big Sur with extFS

### DISCUSSION THREAD

**Discussion by istvanzk** [gitbls/sdm#6] [Q&A] [answered]
Hi,

I was wondering if sdm would work when installed on Mac OS Catalina (or Big Sur) with extFS by Paragon Software to be able to access ext4 file systems?
I'm not sure either if the required systemd functionalities for sdm would/could work under Mac OS.

I understand that some of the settings for the custom RaspiOS img are normally read from the system under which sdm is installed. In my case I'd like to have an option to read these from a custom config file.

Thank you
Istvan

**Comment by gitbls** [comment, marked as answer]
Unfortunately, I don't have a Mac, nor do I know enough about MacOS to answer this definitively. The primary requirements that sdm has on the host OS (MacOS in this case) are systemd-nspawn and ext4. A very quick web search tells me that there _might_ be ways to get both of these working on MacOS. As far as the settings in a config file, you can create a bash script that has the settings you want for such things as Keyboard, Locale, and Timezone so you don't have to set those every time.

I'd be happy to sort out how to use sdm on MacOS if you want to loan me a Mac :) . If that's not possible, I'm happy to work with you on trying to get it working, but you'll need to do most of the footwork.

**Reply by istvanzk** [reply]
Thank you Benn!

Yes, I'm definitely willing to put some work into this, to see if possible to make it work on Mac OS. I'll need to look into the details of using systemd-nspawn. The ext4 part should be solved by the extFS for Mac (by Paragon Sw).

I'll post here my further findings and results.

Best
Istvan

**Reply by gitbls** [reply]
You'll need to install qemu-user-static, which should also pull in binfmt-support. Both of these will be needed to systemd-nspawn into the RasPiOS image. Let me know how it goes!

**Reply by istvanzk** [reply]
I have looked into this option and, while qemu is relatively easy to install on MacOS, it seems is more complicated to make systemd work on MacOS. I'll still try this later if the option below does not work out.

For now, I have started experimenting with another option instead: use VirtualBox (https://www.virtualbox.org) on MacOS with a VM running the latest Raspberry Pi Desktop OS. The VM-Raspi has direct access to the ext4 on the USB SD card reader, without the extFS.

Then I have installed sdm in the VM-Raspi without any problems and downloaded the latest Raspberry Pi OS Lite iso.  
To test further, I run:
`sudo /usr/local/sdm/sdm --explore 2020-12-02-raspios-buster-armhf-lite.img`

The switch into the image happens as expected, however I can see a final error message:
`ERROR: ld.so: object '/usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so' from /etc/ld.so.preload cannot be preloaded (cannot open shared object file): ignored.`

Looking this up, first it seems to be an old qemu issue, fixed in 2019: https://github.com/RPi-Distro/pi-gen/issues/268. 
I suspect, in my case the error might actually be because the VM-Raspi is not running on an expected ARM platform, but the Intel platform of my Mac OS. The $PLATFORM is probably not set (is not an env variable) or is incorrect. I don't know how/if this can be set up manually for qemu.

The ld.so error above could be OK to ignore, maybe, but it is thrown too many times (while in --explore). The apt upgrade is actually failing to update the raspberrypi-kernel, which might be related to the ld.so error.

So, there are some open ends to this VM alternative too.

**Reply by gitbls** [reply]
Istvan, thanks for giving it a try in VirtualBox and creating a repro scenario. I have a Debian x86 VM set up, and was able to replicate that exact error. I'll investigate further and see what I can sort out.

**Reply by gitbls** [reply]
Istvan, here's a workaround that should enable you to get a bit further while I investigate more.

- `sudo /usr/local/sdm/sdm --mount IMGfile.IMG`
- Now the IMG will be mounted on your system. 
- `cd /mnt/sdm/etc`
- `sudoedit ld.preload.so` and comment out the one line that's in there
- `exit`

Now you should be able to do an `sdm --explore`. I haven't tried much more than that, it's time to watch the Seahawks :) Please let me know what other issues you find, and I'll look into this preload issue after the Seahawks clobber the 49ers :) :)

**Reply by istvanzk** [reply]
I did the obvious manual fix to the ld.so error, by editing the ld.so.preload file and replace ${PLATFORM} with v7l. 
This does not fix the apt upgrade error for the raspberrypi-kernel, unfortunately.

**Reply by istvanzk** [reply]
Thank you Benn.
I have not seen your reply before posting mine. Your solution seems more general and I'll use it instead of mine.

**Reply by gitbls** [reply]
Game still on, but did a couple of quick tests. One on Debian 10.6.0, still some problems to chase down. But...good news! My server is running Ubuntu Server 20.04 and was able to do an --extend and --explore the image to do an apt update and upgrade. If you're able to set up an Ubuntu VM you'll probably have better luck than with RPD. Ubuntu incorporates more "leading edge" fixes than Debian or RPD.

**Reply by gitbls** [reply]
Confirmed. I was able to do a complete customize and burn running on Ubuntu 20.04.

**Reply by istvanzk** [reply]
Thank you Benn.

This is good.
I'll switch to use Ubuntu then. I tried with Raspberry Pi Desktop OS / Debian  only because it seemed 'lite' and match the use are better. Maybe Ubuntu has also a lite version. I'll check later.

**Reply by gitbls** [reply]
Istvan, I don't know about Ubuntu desktop, but most definitely with Ubuntu Server there is no GUI installed. I used that when I set up my x64 server. Let me know how that goes, and if you happen to find a Lite version of Ubuntu non-server, I'd be very interested in that as well.

**Reply by istvanzk** [reply]
Yes, the ubuntu server (20.4.1) is lite enough for me too :) I have installed it in the VirtualBox, and I can ssh into it from  my Mac OS host. I have also installed the sdm without any visible errors.
I can mount and check the 2020-12-02-raspios-buster-armhf-lite.img with sdm --mount.

However, when I run sdm --explore I do get an error:
`Mounting IMG '/media/sf_Shared/2020-12-02-raspios-buster-armhf-lite.img'
mount: /dev/loop3p2 mounted on /mnt/sdm.
mount: /dev/loop3p1 mounted on /mnt/sdm/boot.
\*nspawn into image '/media/sf_Shared/2020-12-02-raspios-buster-armhf-lite.img'
Spawning container sdm on /mnt/sdm.
Press ^] three times within 1s to kill container.
execv(/bin/bash, /bin/sh) failed: Exec format error
Container sdm failed with error code 1.
umount: /mnt/sdm/boot unmounted
umount: /mnt/sdm/ unmounted`

**Reply by gitbls** [reply]
Did binfmt-support get installed? 

Also, not sure if this affects you, but you might want to pick up sdm V3.21 (only change is sdm itself) as there was an error in device name handling for /dev/mmcblk devices when burning or mounting them.

**Reply by istvanzk** [reply]
Aha. My bad. I thought qemu-user-static is installed by the sdm installation script. 
I have installed it manually with apt, and now --explore is started without errors. I'm using the latest sdm version.

However, when in explore mode, the apt update of the img fails with:
`W: Failed to fetch http://raspbian.raspberrypi.org/raspbian/dists/buster/main/binary-armhf/Packages  Error writing to output file - write (28: No space left on device) [IP: 93.93.128.193 80]`

Probably I'm missing some configuration, I'll look into this later.

**Reply by gitbls** [reply]
Sounds like the IMG file doesn’t have enough free space in it to hold all the updates. Try doing `sdm — extend —xmb 2048 name-of-img-file.IMG.  sdm used to do the extend automatically, but I kept running into cases where it extended when I didn’t want it to, so I removed that from the customization step. Now you need to explicitly extend the image. If you don’t...well, you’ve seen the result. ;(

**Reply by istvanzk** [reply]
Super. With the extended option it works like a charm! Thank you Benn!

**Reply by gitbls** [reply]
Great news! I’ll be interested in your thoughts and feedback after you’ve used it for a while. What’s good, and perhaps more importantly, what shortcomings or areas where improvements could be helpful did you run in to?

And (shameless plug) if you really like it, it would be awesome if you could do a post in the Raspberry Pi forums to help expose others to the benefits of using sdm. Not necessary of course, and I’ll still work with you on any future issues whether you do or not. ;)

**Reply by istvanzk** [reply]
Sure Benn. 
I'll provide my findings and ideas, and also advertise the sdm once I got to use it more.
One first thing which you could consider including in the installation script is the installation of the qemu-user-static.

**Reply by gitbls** [reply]
Thanks, Istvan. I've been thinking about the installation script since revisiting this with you yesterday, and I've decided that I'll a) check the OS that it's being installed on, and b) If not Raspbian/RasPiOS then install qemu-user-static and binfmt-support, and c) warn if the OS isn't raspios or ubuntu that it might not work.

**Reply by gitbls** [reply]
EZsdmInstaller is now OS-aware per my previous note.

**Comment by gitbls** [comment]
One other thought. If you have a Pi, you can use the Pi itself to bootstrap using sdm:

1. Install RasPiOS or RasPiOS Lite onto an SD card and boot it on your Pi.
2. Install sdm onto your running SD Card
3. Use sdm to create an image just the way you  like it
4. Burn that image onto an SD Card. You'll need a second SD Card and a USB SD Card reader to complete this step
5. Boot the SD Card that you just burned, and you're ready to go. No need to use the standard unmodified RasPiOS ever again.

**Reply by istvanzk** [reply]
Thank you Benn

Indeed this was also the first option I thought about. I have only one Pi (actually also a very old laptop running PiOS) which I could use in this way, the others are more 'embedded' into projects, etc.
Let's see how the Mac OS option works out and I keep this as the fall-back solution.

Best
Istvan
